Email Pro download links

jessjenk opened the pull request:
Use Cloud Functions to process Pro download requests and email users with the results. The [README](README.md) describes how this works, so I'll limit this PR description to a set of TODOs. This PR is already pretty big, so I'd prefer to do them in follow-up PRs. There's some inline questions and clarifications, too.

TODO:
* `inv` tasks for deploying
* secure the endpoint so it can only be called by the Pro Frontend and gcloud users
* error timeout/handling
* Sentry
* Datadog
* CircleCI
* Deploy to `aclima-lab`

jessjenk commented on bq_query.py:
```diff
@@ -0,0 +1,236 @@
+
+from bq_client import BQQueryParams
+
+BQ_PRODUCT_HYPERLOCAL_TABLE = "product_hyperlocal_v1"
```
I could extend the data written in `contract_tile_mapping.json` and pull it from there. Thoughts?

tdlangland replied:
Ohh interesting. I think that would work. 

In case it helps reduce repetition, the tables should be common across a product (e.g.`aclima-pro`) vs needing to be specified per contract. Whether that is easier or harder to write/parse is a different question.

jessjenk commented on exporter.py:
```diff
@@ -0,0 +1,103 @@
+    @property
+    def bq_client(self) -> BQClient:
+        if not self._bq_client:
+            self._bq_client = BQClient()
```
It's defaulting to the project that the Cloud Function is deployed in.

dzhurley commented on main.py:
```diff
@@ -0,0 +1,148 @@
+
+    now = pendulum.now("utc")
+    body["timestamp"] = now.isoformat()
+    now_filepath = now.format("YYYYMMDD/HH:mm")
```
got it. i don't think it's too draconian, i'd like to see measurements over time to inform us if/when this might need to change.

dzhurley replied:
i don't think it's too draconian, i'm in for this for now and watching measurements to see if the limits are getting hit often later.

jessjenk replied:
i'll work on stats in another PR

dzhurley commented on main.py:
```diff
@@ -0,0 +1,148 @@
+            bq_export_table_name = bq_export_table_name.replace("-", "_")
+            # CSV result is written to same folder that `request.json` was pulled from.
+            gcs_export_blob_name = (
+                f"{os.path.dirname(blob.name)}/{export_request['type']}_{idx}.csv.gz"
```
that's a really good point about same data/modality constraints, as that could potentially happen for separate geometries. a short name would be awesome for these files if possible, although we'd have to be okay with having the same short name for potentially different contracts (e.g. `san-diego_quarterly_2019-Q2` and `san-diego_quarterly_2019-Q3`). modality isn't as high a priority as i first thought with filename too, especially if there's a future where multiple modalities can be included in one file.

jessjenk replied:
Multiple modalities are now included in a single file, so I dropped it from the filename.

dzhurley commented on send_email.py:
```diff
@@ -0,0 +1,76 @@
+        ]
+
+        msg = EmailMessage()
+        msg["Subject"] = "Aclima Pro Download"
```
if we're alright with duped `region_name` values like `san-diego_quarterly_2019-Q2` and `san-diego_quarterly_2019-Q3` both being `san-diego`, then i'm in for that.

jessjenk commented:
PR updated:
* Enforce that the request only requests one export for each of `hyperlocal`, `segment`, `stationary`
* Changed the downloaded file name format to include region and modality. E.g., `aclima_pro_san_mateo_pm_25_hyperlocal.csv.gz`
* Updated email subject and body to mention region download is for. Added modality to the link hypertext (see screenshot)
* ... and I think addressed all the other stuff, too.

<img width="800" alt="Screen Shot 2019-08-30 at 2 01 00 PM" src="https://user-images.githubusercontent.com/1255822/64051028-a0e7f100-cb2e-11e9-9bcb-4b34252955fe.png">

tdlangland commented on bq_query.py:
```diff
@@ -0,0 +1,236 @@
+        where_clauses.extend(minmax_where)
+        query_params.extend(minmax_params)
+
+        segment_subquery = f"""
```
Wondering if it might be worthwhile using some context about contracts to avoid doing full table scans for these.

Right now they are scanning >50GB, which can be cut to < 1GB if we look for aggregate data on just the last day of the contract (if it's over) or 'yesterday' UTC (could expand this to a couple days in case the latest aggregation job fails).

jessjenk replied:
This was tricky! I ended up issuing a separate query to get the `MAX(agg_end_time)` for a contract using its end time information from the config. I tried to incorporate it in a single query, but even though the subqueries were optimized, it couldn't seem to optimize overall.

dzhurley commented on examples/export_request.json:
```diff
@@ -0,0 +1,30 @@
+{
+    "email": "jessica.jenkins@aclimalabs.com",
+    "contract": "san-diego_quarterly_2019-Q2",
```
what kind of lift would pushing this to live in each export (next to modality, geometry, etc) be? i ask because we've got some customers in the san mateo/richmond area that are granted access to the entire district stationary set. to keep the correct start/end time constraints on the queries (e.g. richmond is quarterly but district is annual), we need to differentiate them.

jessjenk replied:
That's fine, but can we keep the most specific contract at the outer scope as well for tracking the request/response per-contract?

dzhurley replied:
That works for me, by specific do you mean the one covering the least area?

jessjenk replied:
The question I'm asking is "which contract should this user request be tracked for?" Whatever that is should be at the top-level.

dzhurley replied:
gotcha, that makes sense to me!

jessjenk commented on bq_query.py:
```diff
@@ -0,0 +1,236 @@
+        query_params.extend(minmax_params)
+
+        segment_subquery = f"""
+        SELECT DISTINCT segment_id, contract, data_version
```
I ended up doing this, with a lot more columns in the `SELECT`/`GROUP BY`

jessjenk commented:
This is ready for another look! Updates since the last iteration:
* multiple modalities can be specified in each export request. See `examples/export_request.json`.
* contract can optionally be specified on each export request type. E.g., the stationary request can use a different contract.
* optimized segment query
* some tests (but I still need to set up CircleCI)
* removed `representativeness`
